File: src/algorithms/python/Maximum_Product_Subarray.py
# -*- coding: utf-8 -*-


class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def maxProduct(self, nums):
        length = len(nums)
        tmp, ret = 1, max(nums)
        for i in range(length):
            tmp *= nums[i]
            ret = max(ret, tmp)
            if tmp == 0:
                tmp = 1
        tmp = 1
        for i in range(length - 1, -1, -1):
            tmp *= nums[i]
            ret = max(ret, tmp)
            if tmp == 0:
                tmp = 1
        return ret

if __name__ == "__main__":
    solution = Solution()
    print solution.maxProduct([2, 3, -2, 4])

File: src/algorithms/cpp/Maximum_Product_Subarray.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int ret = -0xffffff, tmp = 1;
        for (auto &num: nums)    get_max(ret, num);
        for (int i = 0; i < SZ(nums); ++ i) {
            tmp *= nums[i];
            get_max(ret, tmp);
            if (tmp == 0)   tmp = 1;
        }
        tmp = 1;
        for (int i = SZ(nums) - 1; i >= 0; -- i) {
            tmp *= nums[i];
            get_max(ret, tmp);
            if (tmp == 0)   tmp = 1;
        }
        return ret;
    }
};

int main() {
    Solution solution = Solution();
    vector <int> input_v = {-2};
    cout << solution.maxProduct(input_v) << endl;
    return 0;
}
